bundles/org.openhab.binding.openthermgateway/src/main/java/org/openhab/binding/openthermgateway/handler/OpenThermGatewayHandler.java

```diff
@@ -0,0 +1,238 @@
+
+        // retry connection if disconnect is not explicitly requested
+        if (!explicitDisconnect && config.connectionRetryInterval > 0) {
+            scheduler.schedule(() -> {
+                if (connector != null && !connecting && !connector.isConnected()) {
+                    connect();
+                }
+            }, config.connectionRetryInterval, TimeUnit.SECONDS);
+        }    
+    }
+
+    @Override
+    public void receiveMessage(Message message) {
+        if (DataItemGroup.dataItemGroups.containsKey(message.getID())) {
+            DataItem[] dataItems = DataItemGroup.dataItemGroups.get(message.getID());
+
+            for (DataItem dataItem: dataItems) {
+                String channelId = dataItem.getSubject();
+
+                if (!OpenThermGatewayBindingConstants.SUPPORTED_CHANNEL_IDS.contains(channelId)) {
+                    continue;
+                }
+
+                State state = null;
+
+                switch (dataItem.getDataType()) {
+                    case FLAGS:
+                        state = OnOffType.from(message.getBit(dataItem.getByteType(), dataItem.getBitPos()));
+                        break;
+                    case UINT8:
+                    case UINT16:
+                        state = new DecimalType(message.getUInt(dataItem.getByteType()));
+                        break;
+                    case INT8:
+                    case INT16:
+                        state = new DecimalType(message.getInt(dataItem.getByteType()));
+                        break;
+                    case FLOAT:
+                        float value = message.getFloat();
+                        @Nullable Unit<?> unit = dataItem.getUnit();
+                        state = (unit == null) ? new DecimalType(value) : new QuantityType<>(value, unit);
+                        break;
+                    case DOWTOD:
+                        break;
+                }
+          
+                if (state != null) {
+                    logger.debug("Received update for channel '{}': {}", channelId, state.toFullString());
+                    updateState(channelId, state);
+                }
+            }
+        }
+    }
+
+    @Override
+    public void handleRemoval() {
+        logger.debug("Removing OpenTherm Gateway handler");
+        disconnect();
+        super.handleRemoval();
+    }
+
+    @Override
+    public void dispose() {
+        disconnect();
+        super.dispose();
+    }
+
+    private boolean checkConnection() {
+        if (connector != null && connector.isConnected()) {
+            return true;
+        }
+
+        return connect();
+    }
+
+    private synchronized boolean connect() {
+        disconnect();
+
+        logger.info("Starting OpenTherm Gateway connector");
+
+        explicitDisconnect = false;
+
+        connector = new OpenThermGatewaySocketConnector(this, config.ipaddress, config.port);
+        new Thread(connector).start();
```

Comment:
From what I understand, the scheduler is used to execute a particular task at a given time, optionally recurring at an interval. This would work perfectly for, let say, reading data from some REST API. This thread however is used to have a constant loop listening for incoming data on the TCP socket, for which (again from what I understand) the Scheduler would not be suitable.

I have asked questions about this on the forum, but didn't get a helpful answer. I also looked at at implementations from other bindings that have a similar character (ie. the constant loop) and there a separate thread is also used. Also, previous reviews from Hilbrand and J-N-K (amongst others) have so far not commented on using a thread in this particular situation.

Anyway, if you feel that this should really be solved using a Scheduler, I will try to find some time to furter investigate the options.

Reply:
Alright I see your point. Then at the very least you need to make sure that the created thread is named and is daemon. Make sure the thread name includes the name of the binding and its purpose.

Reply:
Thanks again! Added the name and made the thread daemon